File: src/block/extrinsics/assurances.rs
// src/extrinsics/assurances.rs

use crate::bitsequence::BitSequence;
use crate::types::{Ed25519Signature, ValidatorIndex}; // assuming you have a BitSequence type

#[derive(Debug, Clone)]
pub struct Assurance {
    /// Must be equal to the ParentHash field of the header.
    pub parent_hash: [u8; 32],
    pub core_availability_contributions: BitSequence,
    pub validator_index: ValidatorIndex,
    pub signature: Ed25519Signature,
}

pub type Assurances = Vec<Assurance>;

File: go/block/extrinsics/assurances.go
package extrinsics

import (
	"github.com/ascrivener/jam/bitsequence"
	"github.com/ascrivener/jam/constants"
	"github.com/ascrivener/jam/types"
)

// todo: validate Assurances 11.2.1
type Assurances []Assurance

type Assurance struct {
	ParentHash                    [32]byte                // a
	CoreAvailabilityContributions bitsequence.CoreBitMask // f
	ValidatorIndex                types.ValidatorIndex    // v
	Signature                     types.Ed25519Signature  // s
}

func (a Assurances) HasValidatorIndex(vIndex types.ValidatorIndex) bool {
	for _, assurance := range a {
		if assurance.ValidatorIndex == vIndex {
			return true
		}
	}
	return false
}

func (a Assurances) AvailabilityContributionsForCoreSupermajority(coreIndex types.CoreIndex) bool {
	return a.AvailabilityContributionsForCore(coreIndex) > int(constants.TwoThirdsNumValidators)
}

func (a Assurances) AvailabilityContributionsForCore(coreIndex types.CoreIndex) int {
	sum := 0
	for _, assurance := range a {
		if assurance.CoreAvailabilityContributions.BitAt(int(coreIndex)) {
			sum += 1
		}
	}
	return sum
}
